### ISSUE TITLE
Use Racket error message conventions, at least for srcloc

### ISSUE THREAD

**Issue by greghendershott** [jackfirth/syntax-warn#73]
TLDR please start warnings with `/absolute/path/to/file:<line>:<col>`.

To help work with other tools (including but not limited to Emacs modes like `compilation-mode` and `racket-mode`) it would be good to follow [Racket's error message conventions](https://docs.racket-lang.org/reference/exns.html#%28part._err-msg-conventions%29):

```
‹srcloc›: ‹name›: ‹message›;
 ‹continued-message› ...
  ‹field›: ‹detail›
  ...
```

The key thing (for other tools) is that `‹srcloc›` is the very first thing, and that it follow a predictable format. Although that linked doc doesn't specify `‹srcloc›`, and I can't seem to find doc for that now, in my experience it is:

```
/absolute/path/to/file:<line>:<col>
```

-- and following the tradition that line numbers are 1.. while column numbers are 0.. _shrugs_.

> Note: Some more-recent Racket stuff replaces the `absolute/path/to/` for packages and collections. When it can, [`racket-mode` intentionally avoids that](https://github.com/greghendershott/racket-mode/blob/master/error.rkt#L93-L102). :)

---

If you don't like the esthetics of that being the first line of every error message? In that case I suppose you could have it off by default, but provide a command-line flag that causes it to be included?

**Comment by jackfirth** [comment]
I was sort of keeping `raco setup`, `raco cover`, and `raco test` in mind for output formatting, along with some of the advice [here](http://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout). Specifically, the `raco warn:` and `raco fix:` prefixes seem important because I expect `warn` and `fix` to be used in the shell alongside other commands. I'm not sure how to reconcile this with Racket's error message conventions, which seem to assume the error messages are being displayed in a context that assumes Racket's the only thing running amok.

Would integration with those Emacs modes work if it output something like this?

```
> raco warn -p some-package
raco warn: Checking 3 modules
raco warn: /path/to/module1.rkt
raco warn: /path/to/module2.rkt
raco warn: /path/to/module2.rkt:4:12: some-warning-kind: you did a bad here;
  stx:
    4  (some bad
    5        thing)
  fix:
    4 (not so bad thing)
raco warn: /path/to/module3.rkt
```

This _doesn't_ put srcloc as the first thing, instead using `raco warn:` to mimic how setup, test, and cover function. Would this still work for those tools you mentioned?

**Comment by greghendershott** [comment]
[That advice](http://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout) seems great for things like log files. (And if you were using Racket's `define-logger` it would prepend the logger name.)

But I think the spirit here is more like compilation and search tools. (Errors, warnings, text corresponding to specific file locations. People may want to jump to the locations, quickly and easily.) So I'd suggest leaning that way?  Where the tradition across multiple langs is `^where what`. AFAIK. And from looking at all the regexp's in Emacs' `compilation-error-regexp-alist`.

Of course for `racket-mode` we can customize this. But my guess is that other editors and tools are most likely to "just work" if the warning messages follow the tradition?

---

That link is also interesting because it uses `grep` as an example but the examples are about errors for grep itself (e.g. permissions). But grep's normal output with `-nH` is an example of what I mean :smile::

```
~/src/elisp/racket
$ grep -nH -e racket-repl *.el
racket-bug-report.el:29:  (unless (memq major-mode '(racket-mode racket-repl-mode))
racket-bug-report.el:30:    (user-error "Please run this from a racket-mode or racket-repl-mode buffer."))
racket-collection.el:21:(require 'racket-repl)
racket-common.el:19:;; Things used by both racket-mode and racket-repl-mode
racket-common.el:195:      (eq major-mode 'racket-repl-mode)))
   . . . .
```

And as used by some Emacs commands all those locations will be "linked" so you can click them, and the usual `next-error` command works, and so on.

**Comment by jackfirth** [comment]
Okay, I think I've got an idea to resolve this:
- Warning messages, their source, and their suggested fix are displayed without a `raco warn:` prefix and with racket's standard error message conventions
- Informational messages from `raco warn`, like how many modules are being checked or which module is in the process of being checked, are displayed with a `raco warn:` prefix
- Informational messages can be silenced with a `--quiet` flag, which _only_ displays warning messages/sources/fixes

So output might look like this normally:

```
> raco warn -p some-package
raco warn: Checking 3 modules
raco warn: /path/to/module1.rkt
raco warn: /path/to/module2.rkt
/path/to/module2.rkt:4:12: some-warning-kind: you did a bad here;
  stx:
    4  (some bad
    5        thing)
  fix:
    4 (not so bad thing)
raco warn: /path/to/module3.rkt
```

But with the `--quiet` flag, it would look like this:

```
> raco warn -p --quiet some-package
/path/to/module2.rkt:4:12: some-warning-kind: you did a bad here;
  stx:
    4  (some bad
    5        thing)
  fix:
    4 (not so bad thing)
```

Would that work for `compilation-mode` / `racket-mode`?

**Comment by jackfirth** [comment]
Thinking more, perhaps the informational messages should be logged to `stderr` instead of `stdout` as well. What do you think?

**Comment by greghendershott** [comment]
> - Warning messages, their source, and their suggested fix are displayed without a raco warn: prefix and with racket's standard error message conventions

As far as I can tell this ought to be the only _necessary_ item in order for compilation-mode and friends to "see" the srclocs and do their usual helpful things.

---

The rest are optional:

> - Informational messages from `raco warn`, like how many modules are being checked or which module is in the process of being checked, are displayed with a `raco warn:` prefix

That's fine.

> - Informational messages can be silenced with a `--quiet` flag, which only displays warning messages/sources/fixes

compilation-mode &c will ignore these anyway -- won't match the regexp for errors.

Although it would make the `*compilation*` buffer less noisy, I tend to use `next-error` to skip over the noise and get to the interesting bits, anyway.

(Maybe `--quiet` would be useful for other tools/purposes, I don't know.)

> Thinking more, perhaps the informational messages should be logged to stderr instead of stdout as well.

compilation-mode &c wouldn't be able to distinguish this -- because when Emacs runs a process, both stdin and stderr end up mixed in the buffer.

(Maybe useful for other tools/purposes, I don't know.)

---

I think only the first item -- present warnings with srclocs in the standard Racket format as for errors -- is required for compilation-mode &c.

**Comment by jackfirth** [comment]
I'll do the first part and close this issue once that's completed then. The other parts are conveniences - if I'm checking for warnings on a lot of modules in CI, I don't want a hundred lines of informational messages polluting the build log.
